src/NuGet.Core/NuGet.Protocol/Resources/VulnerabilityInfoResourceV3.cs

```diff
@@ -0,0 +1,299 @@
+                throw new FatalProtocolException(message: ex.Message, innerException: ex);
+            }
+
+            if (data == null)
+            {
+                string message = "Package source '{0}' vulnerability data file '{1}' ({2}) did not successfully deserialize. This indicates a problem with the JSON file contents returned from the server.";
+                throw new FatalProtocolException(message);
+            }
+
+            return data;
+        }
+
+        /// <inheritdoc cref="IVulnerabilityInfoResource.GetVulnerabilityInfoAsync(SourceCacheContext, ILogger, CancellationToken)"/>
+        public async Task<GetVulnerabilityInfoResult> GetVulnerabilityInfoAsync(SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
+        {
+            List<Exception>? exceptions = null;
+            List<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>>? knownVulnerabilities = null;
+
+            try
+            {
+                IReadOnlyList<V3VulnerabilityIndexEntry> indexEntries = await GetVulnerabilityFilesAsync(cacheContext, logger, cancellationToken);
+
+                if (indexEntries.Count == 0)
+                {
+                    knownVulnerabilities = new();
+                    return new GetVulnerabilityInfoResult(knownVulnerabilities, ToAggregateException(exceptions));
+                }
+
+                var tasks = new Task<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>>[indexEntries.Count];
+                for (int i = 0; i < tasks.Length; i++)
+                {
+                    V3VulnerabilityIndexEntry indexEntry = indexEntries[i];
+                    tasks[i] = GetVulnerabilityDataAsync(indexEntry, cacheContext, logger, cancellationToken);
+                }
+
+                indexEntries = GetValidIndexEntries(indexEntries, ref exceptions);
+
+                await Task.WhenAll(tasks);
+
+                for (int i = 0; i < tasks.Length; i++)
+                {
+                    try
+                    {
+                        IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>? data = await tasks[i];
+                        if (data != null)
+                        {
+                            if (knownVulnerabilities == null)
+                            {
+                                knownVulnerabilities = new(indexEntries.Count);
+                            }
+                            knownVulnerabilities.Add(data);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        AddException(ref exceptions, ex);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                AddException(ref exceptions, ex);
+            }
+
+            GetVulnerabilityInfoResult result = new(knownVulnerabilities, ToAggregateException(exceptions));
+            return result;
```

Comment:
Okay now I have another thought, what if this method just threw the exception at the end and the `GetVulnerabilityInfoResult` doesn't have a property indicating the exceptions.  Wouldn't that be a bit more aligned with what exceptions are for?  The caller would use `try...catch` instead of getting back an object and checking a property.  What do you think?

Reply:
As explained in [another comment I just posted](https://github.com/NuGet/NuGet.Client/pull/5115#discussion_r1159487887), I want this to degrade gracefully when errors occur.  If a server reports 5 pages of vulnerability data, but only one of them has an error, I think it's still useful (and important) to use the 4 good pages of data to check if restored packages have known vulnerabilities.

Reply:
Okay I obviously missed the design of returning potentially incomplete results.